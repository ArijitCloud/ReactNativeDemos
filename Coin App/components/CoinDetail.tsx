import React, { } from 'react';
import { Platform, StyleSheet, Text, useWindowDimensions, View } from 'react-native';
import { CoinItem } from '../common/Models';
import RenderHtml from 'react-native-render-html';
import { limitLargeText } from '../common/Utility';

/**
 * CoinDetail component takes detail of single coin item and displays it.
 * Some section of the code in this component commented out with explanation
 */
export const CoinDetail = ({ coin }: { coin: CoinItem }) => {

    // Limit large text for optimal view in specific device - numberoflines props control breaking for specific device
    // So this measure is still required even if using numberoflines props of Text
    let descriptionText = limitLargeText(coin.Description, 1000);
    descriptionText += Platform.OS === 'web' ? '' : '...';

    return (
        <View style={styles.container} >
            <Text style={styles.subSectionContainer}>
                <Text style={styles.subSectionTitle}>Name: </Text>
                <Text style={styles.subSectionValue}>{coin.Name}</Text>
            </Text>
            <Text style={styles.subSectionContainer}>
                <Text style={styles.subSectionTitle}>Symbol: </Text>
                <Text style={styles.subSectionValue}>{coin.Symbol}</Text>
            </Text>
            <Text style={styles.subSectionContainer}>
                <Text style={styles.subSectionTitle}>Hashing Algorithm: </Text>
                <Text style={styles.subSectionValue}>{coin.HashingAlgorithm ?? 'No hashing algorithm value found for this coin!'}</Text>
            </Text>
            <Text style={styles.subSectionContainer}>
                <Text style={styles.subSectionTitle}>Description: </Text>
                {/* Render html causing  device specific issue. But works well in web */}
                {/* <View style={{ flex: 1, padding: 8 }}>
                    <RenderHtml contentWidth={useWindowDimensions().width} source={{ html: descriptionText }} />
                </View> */}
                <Text style={styles.subSectionValue} numberOfLines={2}>
                    {descriptionText}
                </Text>

            </Text>
            <Text style={styles.subSectionContainer}>
                <Text style={styles.subSectionTitle}>Market Cap (EUR): </Text>
                <Text style={styles.subSectionValue}>{coin.MarketCap}</Text>
            </Text>
            <Text style={styles.subSectionContainer}>
                <Text style={styles.subSectionTitle}>Homepage: </Text>
                <Text style={styles.subSectionValue}>{coin.Homepage}</Text>
            </Text>
            <Text style={styles.subSectionContainer}>
                <Text style={styles.subSectionTitle}>Genesis Date: </Text>
                <Text style={styles.subSectionValue}>{coin.GenesisDate ?? 'No genesis date found for this coin!'}</Text>
            </Text>
        </View >
    );
}

/**
 * CoinDetail component styles using css in js definition
 */
const styles = StyleSheet.create({
    container: {
        paddingTop: 40
    },
    separator: {
        marginVertical: 30,
        height: 1,
        width: '100%',
        backgroundColor: '#CED0CE'
    },
    header: {
        fontSize: 20
    },
    subSectionContainer: {
        marginBottom: 8,
        flex: 1

    },
    subSectionTitle: {
        fontSize: 15,
        fontWeight: '600',
    },
    subSectionValue: {
        fontSize: 14
    }
});
