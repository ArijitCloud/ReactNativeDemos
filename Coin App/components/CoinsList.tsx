import React, { useState, useEffect } from 'react';
import { Button, StyleSheet, Image, Text, View, Pressable, Modal, ActivityIndicator, ScrollView } from 'react-native';
import { CoinItem, MarketList } from '../common/Models';
import { formatCoinDetail, formatMarketList } from '../common/Utility';
import QueryParameters from '../constants/QueryParameters';
import { getCoinDetailsAsync, getCoinMarketsAsync } from '../services/Coingecko.service';
import { DataTable } from 'react-native-paper';
import { CoinDetail } from './CoinDetail';

/**
 * CoinList component shows a tabular view for all the coins
 */
export const CoinsList = () => {
    const [coinmarketList, setCoinMarketList] = useState<MarketList[]>();
    const [coinData, setCoinData] = useState<CoinItem>();
    const [loadingCoin, setLoadingCoin] = useState<boolean>(false);
    const [showDetail, setShowDetail] = useState<boolean>(false);
    const { currency, marketsOrderby, perPage } = QueryParameters;

    /**
     * Load list of coin data from api only once when coin list component is loaded first time
     */
    useEffect(() => {
        const queryString = `?vs_currency=${currency}&order=${marketsOrderby}&per_page=${perPage}&page=1&sparkline=false`;
        getCoinMarketsAsync(queryString).then((d) => {
            //Format the raw data and set the state
            const formattedData = formatMarketList(d);
            setCoinMarketList(formattedData);
        })
    }, []);

    /**
     * On click of a row in tabular view query api with coin id
     */
    const onRowClick = (id: string) => {
        //Set modal visibilty and loading progress spinner visibility
        setShowDetail(true);
        setLoadingCoin(true);

        //Query api
        const queryString = `?localization=false&tickers=false&market_data=true&community_data=false&developer_data=false&sparkline=false`;
        getCoinDetailsAsync(id, queryString).then(d => {
            //Format the raw data and set the state
            const formattedData = formatCoinDetail(d);
            setCoinData(formattedData);
            //Set hide loading progress spinner
            setLoadingCoin(false);
        })
    };

    return (
        <View style={styles.container}>
            <Text style={styles.header}>{'Coins List'}</Text>
            <View style={styles.separator}></View>
            {coinmarketList && coinmarketList.length > 0 &&
                <DataTable key="datatable">
                    <DataTable.Header>
                        <DataTable.Title key='nameheader'>Name</DataTable.Title>
                        <DataTable.Title key='symbolheader'>Symbol</DataTable.Title>
                        <DataTable.Title numeric key='priceheader'>Current Price</DataTable.Title>
                        <DataTable.Title numeric key='highheader'>High (24h)</DataTable.Title>
                        <DataTable.Title numeric key='lowheader'>Low (24h)</DataTable.Title>
                    </DataTable.Header>
                    {coinmarketList.map((coin, index) => (
                        <View key={'viewcontainer' + index}>
                            <Pressable onPress={() => onRowClick(coin.Id)} key={'pressable' + index}>
                                <DataTable.Row key={'row' + index}>
                                    <DataTable.Cell style={styles.cellStyle}>
                                        <Image style={styles.image} source={{ uri: coin.Image }} />
                                        {coin.Name}
                                    </DataTable.Cell>
                                    <DataTable.Cell>{coin.Symbol}</DataTable.Cell>
                                    <DataTable.Cell numeric>{coin.CurrentPrice}</DataTable.Cell>
                                    <DataTable.Cell numeric>{coin.High}</DataTable.Cell>
                                    <DataTable.Cell numeric>{coin.Low}</DataTable.Cell>
                                </DataTable.Row>
                            </Pressable>
                            {/* Below section implement modal view on click of each row */}
                            {showDetail &&
                                <View style={styles.centeredView} key={'modal' + index}>
                                    <Modal animationType="fade" visible={showDetail} key={coin.Id}>
                                        <View style={styles.centeredView}>
                                            <View style={styles.modalView}>
                                                {
                                                    !loadingCoin && coinData && <ScrollView style={{}}>
                                                        <CoinDetail coin={coinData} />
                                                        <View style={styles.buttonContainer}>
                                                            <Button title='Close' key={'btnClose'}
                                                                onPress={() => setShowDetail(false)}>Close</Button>
                                                        </View>

                                                    </ScrollView>
                                                }
                                                {
                                                    loadingCoin && <ActivityIndicator size="large" color="#00ff00" />
                                                }

                                            </View>
                                        </View>
                                    </Modal>
                                </View>
                            }
                        </View>
                    ))
                    }
                </DataTable>
            }
        </View>
    );
}

/**
 * CoinList styles css in js definitions
 */
const styles = StyleSheet.create({
    container: {
        alignItems: 'center',
        justifyContent: 'center',
        paddingTop: 40
    },
    separator: {
        marginVertical: 30,
        height: 1,
        width: '100%',
        backgroundColor: '#CED0CE'
    },
    header: {
        fontSize: 20
    },
    image: {
        height: 20,
        width: 20,
        marginRight: 4,
        flex: 1,
    },
    centeredView: {
        flex: 1,
        justifyContent: "center",
        alignItems: "center",
        marginTop: 22,
        marginLeft: '10%',
        maxWidth: '80%'
    },
    cellStyle: {
        marginRight: 4
    },
    buttonContainer: {
        alignItems: "center",
    },
    modalView: {
        margin: 20,
        backgroundColor: "white",
        borderRadius: 20,
        padding: 20,
        alignItems: "center",
        shadowColor: "#000",
        shadowOffset: {
            width: 0,
            height: 2
        },
        shadowOpacity: 0.25,
        shadowRadius: 4,
        elevation: 5,
    },
});
