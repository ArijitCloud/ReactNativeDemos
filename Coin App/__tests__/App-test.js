import React from 'react';
import renderer from 'react-test-renderer';
import { App } from '../App';
import { CoinsList } from '../components/CoinsList';
import { CoinDetail } from '../components/CoinDetail';


describe("App snapshot verification", () => {

    /**
     * Method called to verify snapshot for App component
     */
    // test('App renders correctly', () => {
    //     const tree = renderer.create(<App />).toJSON();
    //     expect(tree).toMatchSnapshot();
    // });

    /**
     * Method called to verify snapshot for CoinsList component
     * TODO: issues with fetch - Add Mock data
     */
    test('Coins list renders correctly', () => {
        const tree = renderer.create(<CoinsList />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    /**
    * Method called to verify snapshot for CoinDetail component
    * TODO: issues with fetch - Add Mock data
    */
    // test('Coin Detail renders correctly', () => {
    //     const tree = renderer.create(<CoinDetail />).toJSON();
    //     expect(tree).toMatchSnapshot();
    // });
});