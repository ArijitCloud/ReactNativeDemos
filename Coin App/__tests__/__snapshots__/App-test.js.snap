// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot verification Coins list renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "paddingTop": 40,
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 20,
      }
    }
  >
    Coins List
  </Text>
  <View
    style={
      Object {
        "backgroundColor": "#CED0CE",
        "height": 1,
        "marginVertical": 30,
        "width": "100%",
      }
    }
  />
</View>
`;
